opal/mca/btl/ofi/btl_ofi_module.c

```diff
@@ -0,0 +1,296 @@
+ * Copyright (c) 2004-2005 The Regents of the University of California.
+ *                         All rights reserved.
+ * Copyright (c) 2014-2018 Los Alamos National Security, LLC. All rights
+ *                         reserved.
+ * Copyright (c) 2018      Intel, Inc, All rights reserved
+ *
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "opal_config.h"
+#include <string.h>
+#include "opal/class/opal_bitmap.h"
+#include "opal/mca/btl/btl.h"
+#include "opal/datatype/opal_convertor.h"
+#include "opal/mca/mpool/base/base.h"
+#include "opal/mca/mpool/mpool.h"
+
+#include "btl_ofi.h"
+#include "btl_ofi_endpoint.h"
+
+static int mca_btl_ofi_add_procs (mca_btl_base_module_t *btl,
+                                  size_t nprocs, opal_proc_t **opal_procs,
+                                  mca_btl_base_endpoint_t **peers,
+                                  opal_bitmap_t *reachable)
+{
+    int rc;
+    int count;
+    char *ep_name = NULL;
+    size_t namelen = mca_btl_ofi_component.namelen;
+
+    mca_btl_ofi_module_t *ofi_btl = (mca_btl_ofi_module_t *) btl;
+
+    for (size_t i = 0 ; i < nprocs ; ++i) {
+        peers[i] = mca_btl_ofi_endpoint_create (opal_procs[i], ofi_btl->ofi_endpoint);
+        if (OPAL_UNLIKELY(NULL == peers[i])) {
+            return OPAL_ERR_OUT_OF_RESOURCE;
+        }
+
+        OPAL_MODEX_RECV(rc, &mca_btl_ofi_component.super.btl_version,
+                        &peers[i]->ep_proc->proc_name, (void **)&ep_name, &namelen);
+        if (OPAL_SUCCESS != rc) {
+            BTL_ERROR(("error receiving modex"));
+            MCA_BTL_OFI_ABORT();
+        }
+
+        /* get peer fi_addr */
+        count = fi_av_insert(ofi_btl->av,      /* Address vector to insert */
+                             ep_name,          /* peer name */
+                             1,                /* amount to insert */
+                             &peers[i]->peer_addr, /* return peer address here */
+                             0,                /* flags */
+                             NULL);            /* context */
+
+        /* if succeed, add this proc and mark reachable */
+        if (count == 1) { /* we inserted 1 address. */
+            opal_list_append (&ofi_btl->endpoints, &peers[i]->super);
+            opal_bitmap_set_bit(reachable, i);
+        } else {
+            BTL_VERBOSE(("fi_av_insert failed with rc = %d", count));
+            MCA_BTL_OFI_ABORT();
+        }
+    }
+
+    return OPAL_SUCCESS;
+}
+
+static int mca_btl_ofi_del_procs (mca_btl_base_module_t *btl, size_t nprocs,
+                                  opal_proc_t **procs, mca_btl_base_endpoint_t **peers)
+{
+    mca_btl_ofi_module_t *ofi_btl = (mca_btl_ofi_module_t *) btl;
+
+    for (size_t i = 0 ; i < nprocs ; ++i) {
+        if (peers[i]) {
+            opal_list_remove_item (&ofi_btl->endpoints, &peers[i]->super);
+            OBJ_RELEASE(peers[i]);
+        }
+    }
+
+    return OPAL_SUCCESS;
```

Comment:
Is it worth removing the av entry here, too?

Reply:
Why not?
fixed.